### PR TITLE
Order exercise

### PR DESCRIPTION
**Pull request by Kalixone**

### CODE REVIEW

File: src/main/java/mate/academy/springbootintro/dto/OrderRequest.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package mate.academy.springbootintro.dto;
+
+public record OrderRequest(
+        String shippingAddress
```
Comment (nowickimj) [inline src/main/java/mate/academy/springbootintro/dto/OrderRequest.java:7]: No validation here?

File: src/main/java/mate/academy/springbootintro/dto/UpdateOrderStatusRequest.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package mate.academy.springbootintro.dto;
+
+import mate.academy.springbootintro.model.Order;
+
+public record UpdateOrderStatusRequest(
+        Order.Status status
```
Comment (nowickimj) [inline src/main/java/mate/academy/springbootintro/dto/UpdateOrderStatusRequest.java:8]: No validation here?

File: src/main/java/mate/academy/springbootintro/model/Book.java [hunk 3]
```diff
@@ -11,6 +11,7 @@
 import jakarta.persistence.GenerationType;
 import jakarta.persistence.JoinTable;
 import jakarta.persistence.ManyToMany;
+import lombok.Data;
```
Comment (nowickimj) [inline src/main/java/mate/academy/springbootintro/model/Book.java]: You only added import but I don't see you using this annotation here.

File: src/main/java/mate/academy/springbootintro/model/Order.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    @Column(name = "status", nullable = false)
+    private Status status;
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(name = "order_date", nullable = false)
+    private LocalDateTime orderDate;
+    @Column(name = "shipping_address", nullable = false)
+    private String shippingAddress;
+    @Column(name = "order_id", nullable = false)
+    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL)
```
Comment (nowickimj) [inline src/main/java/mate/academy/springbootintro/model/Order.java]: adding `orphanRemoval = true` might be useful here

File: src/main/java/mate/academy/springbootintro/model/Order.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+@Setter
+@Entity
+@EqualsAndHashCode(of = "id")
+@Table(name = "orders")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @JoinColumn(name = "user_id", nullable = false)
+    @ManyToOne
```
Comment (nowickimj) [inline src/main/java/mate/academy/springbootintro/model/Order.java]: Use lazy fetching

File: src/main/java/mate/academy/springbootintro/model/OrderItem.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+@Entity
+@Table(name = "orders_items")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @JoinColumn(name = "order_id", nullable = false)
+    @ManyToOne
+    private Order order;
+    @JoinColumn(name = "book_id", nullable = false)
```
Comment (nowickimj) [inline src/main/java/mate/academy/springbootintro/model/OrderItem.java]: Consider using lazy fetching.

File: src/main/java/mate/academy/springbootintro/service/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    private final OrderRepository orderRepostiory;
+    private final OrderMapper orderMapper;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final OrderItemMapper orderItemMapper;
+    private final CartItemRepository cartItemRepository;
+
+    @Override
+    @Transactional
+    public OrderDto placeOrder(Long userId, OrderRequest request) {
+        User user = userRepository.findById(userId)
```
Comment (nowickimj) [inline src/main/java/mate/academy/springbootintro/service/OrderServiceImpl.java]: Use `getReferenceById` so you don't need additional optional handling. This comment applies to all cases below.

File: src/main/resources/db/changelog/changes/11-create-orders-table.yaml [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+                    referencedColumnNames: id
+                    nullable: false
+              - column:
+                  name: status
+                  type: tinyint
+                  constraints:
+                    nullable: false
+              - column:
+                  name: total
+                  type: decimal
```
Comment (nowickimj) [inline src/main/resources/db/changelog/changes/11-create-orders-table.yaml]: Provide precision here.

File: src/main/resources/db/changelog/changes/12-create-order-items-table.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+                    referencedColumnNames: id
+                    nullable: false
+              - column:
+                  name: quantity
+                  type: integer
+                  constraints:
+                    nullable: false
+              - column:
+                  name: price
+                  type: decimal
```
Comment (nowickimj) [inline src/main/resources/db/changelog/changes/12-create-order-items-table.yaml]: Provide precision here.

File: src/main/resources/db/changelog/changes/11-create-orders-table.yaml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+                  name: user_id
+                  type: bigint
+                  constraints:
+                    foreignKeyName: fk_orders_users
+                    referencedTableName: users
+                    referencedColumnNames: id
+                    nullable: false
+              - column:
+                  name: status
+                  type: tinyint
```
Comment (nowickimj) [inline src/main/resources/db/changelog/changes/11-create-orders-table.yaml]: Why tinyint? Shouldn't you consider using enum or at least varchar here?

**Review by nowickimj — CHANGES_REQUESTED** [review]

File: et --hard 64d14ba
Comment (d1sam) [inline et --hard 64d14ba:1]: please remove this file

File: src/main/java/mate/academy/springbootintro/controller/OrderController.java
Comment (d1sam) [inline src/main/java/mate/academy/springbootintro/controller/OrderController.java:1]: Pagination should be implemented for all endpoints where list is returned

File: src/main/java/mate/academy/springbootintro/model/Book.java [hunk 11]
```diff
@@ -41,7 +43,7 @@ public class Book {
     private String coverImage;
     private String description;
     @Column(name = "categories")
-    @ManyToMany(fetch = FetchType.EAGER)
+    @ManyToMany()
```
Comment (d1sam) [inline src/main/java/mate/academy/springbootintro/model/Book.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/springbootintro/model/Order.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    private Long id;
+    @JoinColumn(name = "user_id", nullable = false)
+    @ManyToOne(fetch = FetchType.LAZY)
+    private User user;
+    @Column(name = "status", nullable = false)
+    private Status status;
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(name = "order_date", nullable = false)
+    private LocalDateTime orderDate;
```
Comment (d1sam) [inline src/main/java/mate/academy/springbootintro/model/Order.java:45]:
```suggestion
    @CreationTimestamp
    @Column(name = "order_date", nullable = false)
    private LocalDateTime orderDate;
```

File: src/main/java/mate/academy/springbootintro/service/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    @Override
+    public List<OrderDto> getUserOrderHistory(Long userId) {
+        User user = userRepository.getReferenceById(userId);
+        List<Order> orders = orderRepostiory.findByUser(user);
+        return orders.stream()
+                .map(orderMapper::toDto)
+                .collect(Collectors.toList());
+    }
+
+    @Override
```
Comment (d1sam) [inline src/main/java/mate/academy/springbootintro/service/OrderServiceImpl.java:64]:
```suggestion
    @Override
    @Transactional
```

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but there are some changes, that should be done according my comments.

File: src/main/java/mate/academy/springbootintro/controller/OrderController.java
Comment (nowickimj) [inline src/main/java/mate/academy/springbootintro/controller/OrderController.java:1]: To apply request body validation, use `@Valid`

File: et --hard HEAD~1q
Comment (nowickimj) [inline et --hard HEAD~1q:1]: What's that change?

**Review by nowickimj — CHANGES_REQUESTED** [review]
One thing missing for proper validation

File: et --hard HEAD~1q
Comment (Kalixone) [reply et --hard HEAD~1q:1]: I had some problems and decided to use git reset --hard HEAD~1 command, but it was a bad idea, I'm still learning how it all works, but it's all good now i think so

Comment (nowickimj) [reply et --hard HEAD~1q:1]: Don't use hard reset in your exerices PRs - we want to keep the track of previous commits.

**Review by nowickimj — APPROVED** [review]

**Pull request merged**